File: membership_card_verifier/packages/verifier/src/stages/stage3.ts
import { keccak256, hkdfSha3256, aes256gcmDecrypt } from "../crypto.js";
import { CardProtocolError } from "../errors.js";
import type {
  RpcProvider,
  IpfsProvider,
  CardDocument,
  VerificationError,
  VerifierConfig,
  ChainLink,
} from "../types.js";

export type { ChainLink };

export interface Stage3Result {
  chain_reaches_trusted_root: boolean | "skipped";
  chain_card_addresses: string[];
  chain: ChainLink[];
  errors: VerificationError[];
}

export async function verifyStage3(
  startCardDoc: CardDocument,
  startCardAddress: string,
  rpc: RpcProvider,
  ipfs: IpfsProvider,
  config: Pick<VerifierConfig, "trustedRoots" | "maxChainDepth">,
  startCardPubkey?: Uint8Array
): Promise<Stage3Result> {
  const trustedRoots = config.trustedRoots ?? [];
  const maxDepth = config.maxChainDepth ?? 64;
  const errors: VerificationError[] = [];
  const chainAddresses: string[] = [startCardAddress];
  const chain: ChainLink[] = [
    {
      card_address: startCardAddress,
      public_key: startCardPubkey ? Buffer.from(startCardPubkey).toString("base64url") : "",
      card_content: startCardDoc,
    },
  ];

  let currentDoc = startCardDoc;
  let currentAddress = startCardAddress;

  for (let depth = 0; depth < maxDepth; depth++) {
    const ancestryPubkeys = currentDoc.ancestry_pubkeys;

    if (ancestryPubkeys.length === 0) {
      // Root base case: current card is (or should be) a trusted root
      const isRoot =
        trustedRoots.includes(currentAddress) ||
        (await rpc.isPolicyAuthorizer(currentAddress));
      return {
        chain_reaches_trusted_root: isRoot,
        chain_card_addresses: chainAddresses,
        chain,
        errors,
      };
    }

    // Check if the next address is already a trusted root before walking
    const nextPubkeyB64 = ancestryPubkeys[0];
    if (!nextPubkeyB64) break;
    const nextPubkeyBytes = new Uint8Array(Buffer.from(nextPubkeyB64, "base64url"));
    const nextAddress = keccak256(nextPubkeyBytes);

    const isNextRoot =
      trustedRoots.includes(nextAddress) ||
      (await rpc.isPolicyAuthorizer(nextAddress));

    if (isNextRoot) {
      chainAddresses.push(nextAddress);
      // Note: the root's CardDocument is not fetched/decrypted here (no new I/O per
      // the plan's constraint), so it is not added to `chain` — only to
      // `chain_card_addresses`, which already tracked addresses-only.
      return {
        chain_reaches_trusted_root: true,
        chain_card_addresses: chainAddresses,
        chain,
        errors,
      };
    }

    // Walk one hop: verify the first ancestry_pubkeys entry matches the expected on-chain address
    const cardEntry = await rpc.getCardEntry(nextAddress);
    if (!cardEntry || !cardEntry.exists) {
      errors.push({ stage: 3, code: "CARD_NOT_FOUND", message: `Ancestor card not found: ${nextAddress}` });
      return { chain_reaches_trusted_root: false, chain_card_addresses: chainAddresses, chain, errors };
    }

    const contentKey = hkdfSha3256(nextPubkeyBytes, "card-content-v1");
    let ancestorDoc: CardDocument;
    try {
      const encrypted = await ipfs.fetch(cardEntry.log_head_cid);
      const decrypted = await aes256gcmDecrypt(contentKey, encrypted);
      ancestorDoc = JSON.parse(new TextDecoder().decode(decrypted)) as CardDocument;
    } catch (e) {
      const code = e instanceof CardProtocolError ? e.code : "DECRYPTION_FAILED";
      errors.push({ stage: 3, code, message: String(e) });
      return { chain_reaches_trusted_root: false, chain_card_addresses: chainAddresses, chain, errors };
    }

    chainAddresses.push(nextAddress);
    chain.push({
      card_address: nextAddress,
      public_key: nextPubkeyB64,
      card_content: ancestorDoc,
    });
    currentDoc = ancestorDoc;
    currentAddress = nextAddress;
  }

  errors.push({
    stage: 3,
    code: "CHAIN_DEPTH_EXCEEDED",
    message: `Chain walk exceeded maxChainDepth (${maxDepth})`,
  });
  return {
    chain_reaches_trusted_root: false,
    chain_card_addresses: chainAddresses,
    chain,
    errors,
  };
}

File: membership_card_verifier/packages/verifier-py/src/membership_card_verifier/stages/stage3.py
import base64
import json
from dataclasses import dataclass
from typing import Any, Optional

from membership_card_verifier.crypto import aes256gcm_decrypt, hkdf_sha3_256, keccak256
from membership_card_verifier.errors import CardProtocolError
from membership_card_verifier.types import ChainLink, IpfsProvider, RpcProvider, VerificationError


def _b64url_decode(s: str) -> bytes:
    padding = "=" * (-len(s) % 4)
    return base64.urlsafe_b64decode(s + padding)


def _b64url_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode("utf-8").rstrip("=")


@dataclass
class Stage3Result:
    chain_reaches_trusted_root: bool
    chain_card_addresses: list[str]
    chain: list[ChainLink]
    errors: list[VerificationError]


async def verify_stage3(
    start_card_doc: dict,
    start_card_address: str,
    rpc: RpcProvider,
    ipfs: IpfsProvider,
    config: Any,
    start_card_pubkey: Optional[bytes] = None,
) -> Stage3Result:
    trusted_roots = config.trusted_roots if config.trusted_roots is not None else []
    max_depth = config.max_chain_depth if config.max_chain_depth is not None else 64
    errors: list[VerificationError] = []
    chain_addresses: list[str] = [start_card_address]
    chain: list[ChainLink] = [
        ChainLink(
            card_address=start_card_address,
            public_key=_b64url_encode(start_card_pubkey) if start_card_pubkey else "",
            card_content=start_card_doc,
        )
    ]

    current_doc = start_card_doc
    current_address = start_card_address

    for depth in range(max_depth):
        ancestry_pubkeys = current_doc.get("ancestry_pubkeys", [])

        if len(ancestry_pubkeys) == 0:
            is_root = (
                current_address in trusted_roots
                or await rpc.is_policy_authorizer(current_address)
            )
            return Stage3Result(
                chain_reaches_trusted_root=is_root,
                chain_card_addresses=chain_addresses,
                chain=chain,
                errors=errors,
            )

        next_pubkey_b64 = ancestry_pubkeys[0]
        if not next_pubkey_b64:
            break

        next_pubkey_bytes = _b64url_decode(next_pubkey_b64)
        next_address = keccak256(next_pubkey_bytes)

        is_next_root = (
            next_address in trusted_roots
            or await rpc.is_policy_authorizer(next_address)
        )

        if is_next_root:
            chain_addresses.append(next_address)
            # Note: the root's CardDocument is not fetched/decrypted here (no new I/O per
            # the plan's constraint), so it is not added to `chain` — only to
            # `chain_card_addresses`, which already tracked addresses-only.
            return Stage3Result(
                chain_reaches_trusted_root=True,
                chain_card_addresses=chain_addresses,
                chain=chain,
                errors=errors,
            )

        card_entry = await rpc.get_card_entry(next_address)
        if not card_entry or not card_entry.exists:
            errors.append(
                VerificationError(
                    stage=3,
                    code="CARD_NOT_FOUND",
                    message=f"Ancestor card not found: {next_address}",
                )
            )
            return Stage3Result(
                chain_reaches_trusted_root=False,
                chain_card_addresses=chain_addresses,
                chain=chain,
                errors=errors,
            )

        content_key = hkdf_sha3_256(next_pubkey_bytes, "card-content-v1")
        try:
            encrypted = await ipfs.fetch(card_entry.log_head_cid)
            decrypted = aes256gcm_decrypt(content_key, encrypted)
            ancestor_doc = json.loads(decrypted.decode("utf-8"))
        except CardProtocolError as e:
            errors.append(
                VerificationError(
                    stage=3,
                    code=e.code,
                    message=str(e),
                )
            )
            return Stage3Result(
                chain_reaches_trusted_root=False,
                chain_card_addresses=chain_addresses,
                chain=chain,
                errors=errors,
            )
        except Exception as e:
            errors.append(
                VerificationError(
                    stage=3,
                    code="DECRYPTION_FAILED",
                    message=str(e),
                )
            )
            return Stage3Result(
                chain_reaches_trusted_root=False,
                chain_card_addresses=chain_addresses,
                chain=chain,
                errors=errors,
            )

        chain_addresses.append(next_address)
        chain.append(
            ChainLink(
                card_address=next_address,
                public_key=next_pubkey_b64,
                card_content=ancestor_doc,
            )
        )
        current_doc = ancestor_doc
        current_address = next_address

    errors.append(
        VerificationError(
            stage=3,
            code="CHAIN_DEPTH_EXCEEDED",
            message=f"Chain walk exceeded maxChainDepth ({max_depth})",
        )
    )
    return Stage3Result(
        chain_reaches_trusted_root=False,
        chain_card_addresses=chain_addresses,
        chain=chain,
        errors=errors,
    )
